hibernate-core/src/main/java/org/hibernate/query/hql/internal/QualifiedJoinPathConsumer.java

```diff
@@ -178,8 +179,11 @@ private static SqmFrom<?, Object> createJoin(
 			boolean isTerminal,
 			boolean allowReuse,
 			SqmCreationState creationState) {
+		final SqmPathSource<?> pathSource = lhs instanceof SqmTreatedPath ?
+				( (SqmTreatedPath<?, ?>) lhs ).getTreatTarget() :
+				lhs.getResolvedModel();
 		//noinspection unchecked
-		final SqmPathSource<Object> subPathSource = (SqmPathSource<Object>) lhs.getReferencedPathSource().getSubPathSource( name );
+		final SqmPathSource<Object> subPathSource = (SqmPathSource<Object>) pathSource.getSubPathSource( name );
```

Comment:
As an alternative to this check we could make `getReferencedPathSource()` return the `treatTarget` entity domain type for all treated `SqmJoin`s, like we already do for `SqmTreatedRoot`.

I've done some testing on this and found that, while it would require changing quite a bit of the treated SQM api classes, this should work too. Open to opinions on what is best.

Reply:
>As an alternative to this check we could make getReferencedPathSource() return the treatTarget entity domain type for all treated SqmJoins, like we already do for SqmTreatedRoot.

That solution sounds more correct to me.

Reply:
I've created a separate PR with that https://github.com/hibernate/hibernate-orm/pull/6704.